### PR TITLE
Make snippets translatable

### LINKED ISSUE #6059: wagtail-localize: make all snippets translatable
Unlike models, snippets are not translatable by default. We need to make small adjustments to each of them:

- [ ] Add the `TranslatableMixin` mixin
- [ ] List all fields that need localization to the `translatable_fields` variable
- [ ] Optionally, add fields to `override_translatable_fields` (e.g. ShareProgress fields need to be added to both variables, so that they can be overriden in the admin, but don’t end up in .po files)
- [ ] Probably do something around `unique_together`

Example of translatable snippet: https://github.com/wagtail/wagtail-localize/blob/fdd813e8a9717b8336f31ba9d04b75976d1d915f/wagtail_localize/test/models.py#L200-L216

### PR DESCRIPTION
**Pull request by KalobTaulien**
Closes #6059 

Steps for testing:
1. Run `inv-new db` to create a new database with the migrations in this PR applied 
2. Go to the Wagtail admin -> Settings -> Locales -> Add new locale 
3. Go to Snippets -> Select one of them -> Change Language -> Add new Snippets

### CODE REVIEW

**Comment by KalobTaulien** [comment]
@TheoChevalier Could I get you to give the fields a review? I wasnt sure exactly which fields should be Translatable/Synchronized and when to use overridable fields. I would love your eyes in this PR, thanks!

**Comment by TheoChevalier** [comment]
> @TheoChevalier Could I get you to give the fields a review? I wasnt sure exactly which fields should be Translatable/Synchronized and when to use overridable fields. I would love your eyes in this PR, thanks!

Sure, I’ll read your code and see if i can spot anything wrong, but heads up that I’m hitting issues when I try to translate a campaign page that contains a snippet (Petition, Donation modal, or Signup snippet). When I remove the snippet from the page, I don’t get any issue

Capture d’écran 2021-02-09 à 14 37 55
Capture d’écran 2021-02-09 à 14 38 12

**Comment by TheoChevalier** [comment]
Also found other issues when translating snippets:

1) Petition & SignUp snippets are broken
  - Go to Snippets > Petition
  - Hover over one snippet
  - Hit Translate, select French
  - Switch to French with the language selector at the top
  - Try editing your snippet
  
Capture d’écran 2021-02-09 à 16 09 15

2) Snippet translation is not automatically started when translating an English page that contains them
  -  Start translating the home page in French for instance, and the AreaFocus snippets should be requested in French as well. This is not happening. This is likely a wagtail-localize issue?
Unless we changed our mind last minute during the dev phase and made it a manual process (I may have forgotten…), this was the intended behavior

 3) Translations from translated snippets are not showing up on the page:
  - Start translating the homepage in French.
  - Go to Snippets > Areas of focus
  - Hover over "Empower Action"
  - Hit "Translate", select French
  - Edit the French snippet with dummy translations, hit save
  - Check /fr, your translations are not showing up.

**Comment by KalobTaulien** [comment]
Thanks for looking into this @TheoChevalier 

For number 3 (the only one I can speak for at the moment) every page is translated using an alias page. The alias pages need to select the french version of the snippet - the page won't automatically update the SnippetChoosers to be the french equivalent. 

I'm going to look into the rest of these right now. Thanks!

**Comment by KalobTaulien** [comment]
> 1) Petition & SignUp snippets are broken
>  - Go to Snippets > Petition
>  - Hover over one snippet
>  - Hit Translate, select French
>  - Switch to French with the language selector at the top
>  - Try editing your snippet

I was able to replicate this. As far as I know, this is a bug in wagtail-localize. I thought maybe it was because the translation was set on the CTA class and Petitions/Signups were inheriting from that model, and it might be, but when I tried to set the translations on the specific models instead of the parent model I ended up with the same result. 

For anyone who ends up watching this thread, this is a demo of what Theo found. 
snippet-error

**Comment by KalobTaulien** [comment]
@TheoChevalier For the Signup/Petition snippets issue, I asked Karl about it and he thinks it might be a bug. It seems to only happen with models that inherit from another model. I opened an issue on wagtail-localize here: https://github.com/wagtail/wagtail-localize/issues/321

**Comment by TheoChevalier** [comment]
> Thanks for looking into this @TheoChevalier
> 
> For number 3 (the only one I can speak for at the moment) every page is translated using an alias page. The alias pages need to select the french version of the snippet - the page won't automatically update the SnippetChoosers to be the french equivalent.
> 
> I'm going to look into the rest of these right now. Thanks!

Hmm, I feel like this is a problem then… Because this was definitely the intent, if you translate the snippet in French, the French page should automatically start using that snippet.
Looks like it’s https://github.com/wagtail/wagtail-localize/issues/250 ?

**Comment by KalobTaulien** [comment]
Updating this ticket before I move it into blocked. 

I tried using the latest version of wagtail-localize for the `signup snippets` (one of the problematic snippets because it inherits from the `CTA` model) and we get the same issue. I'm going to move this to blocked for the time being and will check in again in a few days.

**Review by recepsirin — APPROVED** [review]

**Comment by Pomax** [comment]
@recepsirin why are you approving PRs for projects you're not involved with?

**Comment by KalobTaulien** [comment]
Moving back to blocked. The error persists with the latest work from wagtail localize (master branch) 

image

**Comment by KalobTaulien** [comment]
An idea I have this morning is... to denormalize the CTA app and copy the fields into the two inherited models. Yes, we break a normalization rule, but we also unblock ourselves on this ticket. 

It would require some data migrations, but nothing major. And if we wanted to re-normalize the models when wagtail-localize support multi model inheritance we could do that. 

The trade off is: more upfront work now with a possibility of some "revert" work down the road to unblock this ticket, or we wait until wagtail-localize has this fixed.

**Comment by KalobTaulien** [comment]
Wagtail 2.11.6 patches the problem we were having. Im going to open a PR to bump Wagtail to v2.11.6 from its current version of 2.11.3

**Comment by KalobTaulien** [comment]
`master` Has wagtail 2.11.6. `wagtail-localize` branch does not. I added a PR to bump our wagtail version in the long living wagtail-localize branch. This fixes the problem you were having @TheoChevalier 

@TheoChevalier This is ready for re-review. You'll need to use `inv new-env` with this -> create a new locale -> translate  one of the Signup snippets -> select the new locale and edit the translation. 

If the translations don't save, it's because we have this in the settings, which will be updated when we merge this into the wagtail-localize branch (you can comment this out for now to have rest framework allow incoming requests)
```python
REST_FRAMEWORK = {
    'DEFAULT_PERMISSION_CLASSES': [
        'rest_framework.permissions.DjangoModelPermissionsOrAnonReadOnly',
    ]
}
```

File: network-api/networkapi/wagtailpages/pagemodels/campaigns.py [hunk 1]
```diff
@@ -174,7 +188,13 @@ class Petition(CTA):
         default='Thank you for signing too!',
     )
 
-    override_translatable_fields = [
+    translatable_fields = CTA.translatable_fields + [
+        SynchronizedField('campaign_id'),
+        TranslatableField('comment_requirements'),
+        TranslatableField('checkbox_1'),
+        TranslatableField('checkbox_2'),
+        TranslatableField('share_link_text'),
```
Comment (TheoChevalier) [inline network-api/networkapi/wagtailpages/pagemodels/campaigns.py:196]: I’m hitting this when translating petition snippets, I think both `share_link` and `share_link_text` are not used anymore, and could be safely removed?
I assume these were used before ShareProgress?

Capture d’écran 2021-03-11 à 14 57 59

**Review by TheoChevalier — APPROVED** [review]
Thanks @KalobTaulien, I was able to translate most snippets :) (Just had an issue with the petition ones, added more info in a comment))
Feel free to merge this bad boy as-is, I’ve added comments for minor stuff I’ve caught, but we can tackle these in smaller tickets

File: network-api/networkapi/wagtailpages/pagemodels/campaigns.py [hunk 2]
```diff
@@ -46,11 +48,19 @@ class CTA(models.Model):
         default='mozilla-foundation'
     )
 
+    translatable_fields = [
+        TranslatableField('name'),
+        TranslatableField('header'),
+        TranslatableField('description'),
+        TranslatableField('newsletter'),
```
[the remark this answers is not in the document]
Comment (KalobTaulien) [reply network-api/networkapi/wagtailpages/pagemodels/campaigns.py:55]: @TheoChevalier Which line? GitHub has highlighted four of them 😆

File: network-api/networkapi/wagtailpages/migrations/0039_auto_20210208_2114.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+    # Add one operation for each model to bootstrap here
+    # Note: Only include models that are in the same app!
+    operations = [
+        BootstrapTranslatableModel('wagtailpages.BlogAuthor'),
+        BootstrapTranslatableModel('wagtailpages.BlogPageCategory'),
+        BootstrapTranslatableModel('wagtailpages.CTA'),
+        BootstrapTranslatableModel('wagtailpages.DonationModal'),
+        BootstrapTranslatableModel('wagtailpages.FocusArea'),
+        BootstrapTranslatableModel('wagtailpages.FocusArea'),
```
[the remark this answers is not in the document]
Comment (KalobTaulien) [reply network-api/networkapi/wagtailpages/migrations/0039_auto_20210208_2114.py:19]: Correct! Thank you for spotting this!

File: network-api/networkapi/wagtailpages/pagemodels/campaigns.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (TheoChevalier) [reply network-api/networkapi/wagtailpages/pagemodels/campaigns.py:55]: Ops, the `newsletter` field

File: network-api/networkapi/wagtailpages/pagemodels/campaigns.py (hunk 1 above)
Comment (KalobTaulien) [reply network-api/networkapi/wagtailpages/pagemodels/campaigns.py:196]: Hmm yeah, editable=False suggests it's not being used - at least not in a way that editors would have access to. Ill push up a reomval. Thanks for finding this Theo!

File: network-api/networkapi/wagtailpages/pagemodels/campaigns.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (KalobTaulien) [reply network-api/networkapi/wagtailpages/pagemodels/campaigns.py]: Awesome, thanks! It's removed now. Cheers!

**Pull request merged**